From 2b40c79192e3c86074d21af51774971e19cbd2ab Mon Sep 17 00:00:00 2001
From: Uli Schlachter <psychon@znc.in>
Date: Thu, 29 Mar 2012 17:29:27 +0200
Subject: [PATCH] Make this compile with lua 5.2

Big thanks to Evan McClain <aeroevan@gmail.com> for a preliminary patch on which
I based this! (Sorry, but I like my version more :-P)

Signed-off-by: Uli Schlachter <psychon@znc.in>
---
 configure.ac |    2 +-
 oocairo.c    |   21 +++++++++++++++++++++
 2 files changed, 22 insertions(+), 1 deletions(-)

diff --git a/oocairo.c b/oocairo.c
index ae4847f..688c13e 100644
--- a/oocairo.c
+++ b/oocairo.c
@@ -41,6 +41,18 @@
 #error "This Lua binding requires Cairo version 1.6 or better."
 #endif
 
+#if LUA_VERSION_NUM >= 502
+#define lua_objlen(a, b)      lua_rawlen(a, b)
+
+static int luaL_typerror (lua_State *L, int narg, const char *tname) {
+    const char *msg = lua_pushfstring(L, "%s expected, got %s",
+                                      tname, luaL_typename(L, narg));
+    return luaL_argerror(L, narg, msg);
+}
+#else
+#define lua_rawlen(a, b) do_not_use_lua_rawlen_use_lua_objlen_instead;
+#endif
+
 static const int ENDIANNESS_TEST_VAL = 1;
 #define IS_BIG_ENDIAN (!(*(const char *) &ENDIANNESS_TEST_VAL))
 
@@ -785,6 +797,14 @@ write_chunk_to_fh (void *closure, const unsigned char *buf,
     return CAIRO_STATUS_SUCCESS;
 }
 
+#if LUA_VERSION_NUM >= 502
+static void
+get_gtk_module_function (lua_State *L, const char *name) {
+    /* FIXME: Implement this again */
+    (void) name;
+    luaL_error(L, "cannot access gtk with lua >= 5.2, TODO: Implement this again");
+}
+#else
 static void
 get_gtk_module_function (lua_State *L, const char *name) {
     lua_getfield(L, LUA_GLOBALSINDEX, "gtk");
@@ -797,6 +817,7 @@ get_gtk_module_function (lua_State *L, const char *name) {
                    name);
     lua_remove(L, -2);
 }
+#endif
 
 static int
 push_cairo_status (lua_State *L, cairo_status_t status) {
-- 
1.7.2.5

--- a/configure.ac	2011-05-18 20:18:10.000000000 +0200
+++ b/configure.ac	2013-01-30 09:08:59.978630537 +0100
@@ -18,7 +18,7 @@
 # Lua gets different names in different distros :(
 PKG_PROG_PKG_CONFIG
 AC_MSG_CHECKING([for lua])
-for lua_name in lua lua5.1
+for lua_name in lua lua5.1 lua5.2
 do
   PKG_CHECK_EXISTS([$lua_name], [LUA_NAME=$lua_name ; break])
 done
